build: Treat build warnings as errors

philipphofmann opened the pull request:
## :scroll: Description

Treat build warnings for Objective-C and Swift as errors.

This PR is based on #798.

## :bulb: Motivation and Context

Warnings are there for a reason. Let's not ignore them.

## :green_heart: How did you test it?
CI.

## :pencil: Checklist

<!--- Put an `x` in the boxes that apply -->

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] No breaking changes

## :crystal_ball: Next steps

bruno-garcia commented on Sources/SentryCrash/Recording/Tools/SentryCrashUUIDConversion.h:
```diff
@@ -15,7 +15,7 @@ extern "C" {
  *
  */
-static void
+static inline void
```
Was it a warning that led to adding this?
I wonder why explicilty mark this inline. It's often a compiler choice given the trade off in performance vs binary size

philipphofmann replied:
Yes, this caused the following warning `Unused function 'sentrycrashdl_convertBinaryImageUUID'`. This is because this function is marked as static in a header file so each when you include this header file you get an own internal copy of the function. Which is basically what `inline` does. Thinking about it again it doesn't make sense to inline this function. So I changed it now and defined it as a normal c function. While doing this I discovered that the `*dst` was not null-terminated. I don't understand fully how it worked before. I added tests for it and also tested it on a simulator.

bruno-garcia reviewed (commented):
The warnings here would need to be fixed in order for the build to pass?

philipphofmann commented:
I removed `SentryRequestTests` because it only contained mocking code, no runnable tests, and is written in Objective-C. I would have to add tests for it and fix the warning. I added tasks to Asana to add them in the future.

bruno-garcia reviewed (approved):
:shipit:
